lib/spack/spack/relocate.py

```diff
@@ -59,7 +59,10 @@ def get_existing_elf_rpaths(path_name):
         command = Executable(get_patchelf())
         output = command('--print-rpath', '%s' %
```

Comment:
IMO, There are two things here:

- `patchelf` is only applied on files stored in the `yaml` file with all the `buildinfo` metainformation about the binary and the files that need relocation. This file is built when creating the binaries. At that moment `needs_binary_relocation` decides whether the files have to be added to the `yaml` file.

- During the relocation time, `patchelf` relies on the information provided by the `yaml` files, but if any file was not properly checked before adding it to the list of files to be relocated, then `patchelf` of course fails.

I'm not sure whether a double check at binary creation time and at relocation time is really needed. But at least I would make more robust [`needs_binary_relocation`](https://github.com/gartung/spack/blob/687f7e5f3be44b1621f0fdcad390a3d359796214/lib/spack/spack/relocate.py#L219), to be sure that `patchelf` really has something to relocate. Maybe something like:

```python
def needs_binary_relocation(path_name, os_id=None):
    """
    Check whether the given filetype is a binary that may need relocation.
    """
    retval = False
    
    filetype = get_filetype(filename)
    command = Executable('readelf')     
    output = command('-d', '%s' %
                     os.path.realpath(filename), output=str, err=str)
    if "no dynamic section" in output:
        return False
    if "relocatable" in filetype:
        return False
    if "link to" in filetype:
        return False
    if os_id == 'Darwin':
        return ("Mach-O" in filetype)
    elif os_id == 'Linux':
        return ("ELF" in filetype)
    else:
        tty.die("Relocation not implemented for %s" % os_id)
return retval
```

Reply:
@JavierCVilla I agree with what you suggest, however readelf might not be available and the command = Executable line would fail. I am think of using strings instead. What does strings libPartPropSvc.so.dbg produce?

Reply:
@JavierCVilla It would be easier to make a test if you put the file somehwhere that I can download it.

Reply:
@gartung I just did `spack install gaudi@v28r1` using defaults values (just in case: "build_type": "RelWithDebInfo"). Once installed, those `.so.dbg` files are under the folder `lib`:

```console
$ ls spack/opt/spack/linux-scientificcernslc6-x86_64/gcc-6.2.0/gaudi-v28r1-hash/lib/
Gaudi.components                  libGaudiHive.so                   libGaudiPythonLib.so.dbg
Gaudi.confdb                      libGaudiHive.so.dbg               libGaudiPython.so
GaudiDict.rootmap                 libGaudiJemalloc.so               libGaudiPython.so.dbg
GaudiExamplesDict_rdict.pcm       libGaudiJemalloc.so.dbg           libGaudiSvc.so
...
``` 

Anyway, here it is a [link[(https://cernbox.cern.ch/index.php/s/cPe3D6T9T19RA5W) to both.

Reply:
I started the gaudi build but it failed and I didn't have time to debug why. 

I am adding a check for the install root in the output of strings which should only pick elf files with rpaths added.

Reply:
The `NOTICE` file in Spack's base git directory also appears to generate the specific issue that @JavierCVilla got. I think you can create any text file e.g. with `touch test.txt` and it will fail for a different reason but be sufficient to test the issue.

Reply:
@scheibelp I realized that I can't run patchelf directly in the test environment because get_patchelf tries to build patchelf if it does not find it.

I figured out how to check the return code of patchelf --print-rpath and use that to determine the return type.

Reply:
@gartung the issue with that is that if `command.returncode != 0` then an exception would have been thrown before you got to the point where you check from it (see `Executable.__call__` in `lib/spack/spack/util/executable.py`). I think you are going to have to guard the invocation of `command('--print-rpath'...` with a try/except (specifically for `ProcessError`). To make this unit-testable, you could update this method to take the `patchelf` command as an argument, and the unit test could pass in a mocked-up version that always throws an exception.

Does that make sense and seem agreeable?